### PR TITLE
Reorder index add/upd/del defered

### PR DESCRIPTION
**Pull request by adizaimi**
Reorder index operations (add/upd/del) by table and index value and perform them deferred at the end of the osql processing. This ensures that we avoid deadlocking by touching (locking) pages in an increasing order from all the transactions on master node.

### CODE REVIEW

**Comment by riverszhang89** [comment]
Just so you know scindex, scindex_logicalsc_generated, sc_parallel, sc_parallel_logicalsc_generated keep hanging...

**Comment by adizaimi** [comment]
FTR, restart_node failed with error msg "Failed newver is now NULL but should be 0". Looking at the other tests I get:
```
ssh pnj-comdb2btda2 'find /bb/comdb2a/roborivers/tests  | grep testcase | while read filename; do  grep -s -H "Failed newver is now NULL but should be 0" $filename; done'

/bb/comdb2a/roborivers/tests/b032136b/common/c1/db/logs/restartnode75068.testcase:15:03:22> Failed newver is now NULL but should be 0
/bb/comdb2a/roborivers/tests/b032136b/common/c1/db/logs/scresume75068.testcase:15:04:44> Failed newver is now NULL but should be 0
/bb/comdb2a/roborivers/tests/b032136b/common/c1/db/logs/sctransactional75068.testcase:15:12:55> Failed newver is now NULL but should be 0
/bb/comdb2a/roborivers/tests/d7bc69d2/common/c1/db/logs/restartnode39233.testcase:01:35:40> Failed newver is now NULL but should be 0
/bb/comdb2a/roborivers/tests/d7bc69d2/common/c1/db/logs/scresume39233.testcase:01:36:30> Failed newver is now NULL but should be 0
/bb/comdb2a/roborivers/tests/d7bc69d2/common/c1/db/logs/sctransactional39233.testcase:01:51:47> Failed newver is now NULL but should be 0
/bb/comdb2a/roborivers/tests/5d4cd77b/common/c1/db/logs/restartnode87540.testcase:04:31:43> Failed newver is now NULL but should be 0
/bb/comdb2a/roborivers/tests/5d4cd77b/common/c1/db/logs/scresume87540.testcase:04:32:17> Failed newver is now NULL but should be 0
/bb/comdb2a/roborivers/tests/5d4cd77b/common/c1/db/logs/replayeventlog87540.testcase:04:49:08> Failed newver is now NULL but should be 0
/bb/comdb2a/roborivers/tests/5d4cd77b/common/c1/db/logs/sctransactionalrowlocksgenerated87540.testcase:04:51:08> Failed newver is now NULL but should be 0
/bb/comdb2a/roborivers/tests/5d4cd77b/common/c1/db/logs/sctransactional87540.testcase:04:51:20> Failed newver is now NULL but should be 0
/bb/comdb2a/roborivers/tests/9cdd4bb77/common/c1/db/logs/restartnode11171.testcase:19:50:15> Failed newver is now NULL but should be 0
```
which is to say it happens sometimes and it is not related to this checkin, but worth looking into.

**Comment by adizaimi** [comment]
@@ sysbench/oltp_read_write @@ -21.63%
The reason we think this is worse with reordering on is that when writing we get all locks on the data pages  first (in write mode--a little earlier now), then we get the locks on index pages. Readers will block for a bit longer thus resulting in a bit of delay for this workload in single node. 
Comdb2 is not ment to run in single node, and the clustered version for this test shows no slowdown by this feature--investigating same test in a clustered environment shows no difference between reordering on/off. .

File: bdb/fetch.c [hunk 1]
```diff
@@ -2257,11 +2261,13 @@ static int bdb_fetch_int(int return_dta, int direction, int lookahead,
         }
     }
```
Comment (markhannum) [inline bdb/fetch.c]: should get a readlock on the table either way, no?

File: bdb/fetch.c [hunk 2]
```diff
@@ -735,6 +735,10 @@ static int bdb_fetch_int_ll(
             }
 
             dbp = bdb_state->dbp_data[0][dtafile];
+            extern int gbl_reorder_idx_writes;
+            if (args->for_write && gbl_reorder_idx_writes) {
```
Comment (markhannum) [inline bdb/fetch.c]: Yup- looks correct .. might even be beneficial to do this even if gbl_reorder_idx_writes is not enabled, but I think we intend to keep gbl_reorder_idx_writes as the default ..

File: db/indices.c [hunk 3]
```diff
@@ -15,17 +15,146 @@
...
+    assert(sizeof(dtikey_t) == key2len);
+ 
+    dtikey_t *k1 = (dtikey_t *)key1;
+    dtikey_t *k2 = (dtikey_t *)key2;
+
+    CMP_KEY_MEMBER(k1, k2, usedb);
+    CMP_KEY_MEMBER(k1, k2, ixnum);
+    MEMCMP_KEY_MEMBER(k1, k2, ixkey, sizeof(k1->ixkey));
+    CMP_KEY_MEMBER(k1, k2, type);
+    CMP_KEY_MEMBER(k1, k2, genid);
```
Comment (markhannum) [inline db/indices.c]: should genid be first?

File: db/indices.c [hunk 4]
```diff
@@ -15,17 +15,146 @@
...
+ * called from handle_buf.c to cleanup defered tbl */
+void delete_defered_index_tbl() 
+{
+    if (!defered_index_tbl)
+        return;
+
+    if (defered_index_tbl_cursor)
+        close_defered_index_tbl_cursor();
+
+    delete_constraint_table(defered_index_tbl);
```
Comment (markhannum) [inline db/indices.c]: set defered_index_tbl to null

File: db/indices.c [hunk 5]
```diff
@@ -345,27 +527,50 @@ int upd_record_indices(struct ireq *iq, void *trans, int *opfailcode,
...
+    int rc = 0;
     char *od_dta_tail = NULL;
     int od_tail_len;
     char *od_olddta_tail = NULL;
     int od_oldtail_len;
 
+    void *cur = NULL;
+    dtikey_t delditk = {0}; // will serve as the delete key obj
+    dtikey_t ditk = {0};    // will serve as the add or upd key obj
+    bool reorder = gbl_reorder_idx_writes && iq->usedb->sc_from != iq->usedb &&
```
Comment (markhannum) [inline db/indices.c]: so why sc and constraints?

File: db/indices.c [hunk 6]
```diff
@@ -519,33 +780,88 @@ int upd_record_indices(struct ireq *iq, void *trans, int *opfailcode,
...
-                             od_len, opcode, blkpos, opfailcode, newkey,
-                             od_dta_tail, od_tail_len, do_inline, 0);
-
-                if (iq->debug)
-                    reqprintf(iq, "add_key IX %d RRN %d RC %d", ixnum, rrn, rc);
+                if (reorder) {
+                    //if not datacopy, no need to save od_dta_tail
+                    void *data = NULL;
+                    int datalen = 0;
+                    if (iq->usedb->ix_datacopy[ixnum] != 0) { //is datacopy
```
Comment (markhannum) [inline db/indices.c]: if (od_dta_tail)

File: db/indices.c [hunk 7]
```diff
@@ -911,3 +1256,224 @@ int del_new_record_indices(struct ireq *iq, void *trans,
...
+                    rc = ERR_NOMASTER;
+                }
+                goto done;
+            }
+        }
+        else if (ditk->type == DIT_DEL) {
+            int delrrn = 0;
+            char *tblname = iq->usedb->tablename;
+            struct dbtable *tbl = get_dbtable_by_name(tblname);
+            if (tbl != iq->usedb) abort();
```
Comment (markhannum) [inline db/indices.c]: xxx

File: db/indices.c (hunk 3 above)
Comment (adizaimi) [reply db/indices.c]: Adding genid to end of ixkey when ix is dup. Will try how that works.

File: db/indices.c (hunk 4 above)
Comment (adizaimi) [reply db/indices.c]: Yep

File: db/indices.c (hunk 5 above)
Comment (adizaimi) [reply db/indices.c]: I think i saw issues while schema changing. For now lets not reorder if schemachanging and if this table has constraints.

File: db/indices.c (hunk 6 above)
Comment (adizaimi) [reply db/indices.c]: Changing.

File: db/indices.c (hunk 7 above)
Comment (adizaimi) [reply db/indices.c]: turning it into assert.

**Comment by adizaimi** [comment]
Closing this because it is not being picked up by roborivers. 
Fixed the issues discussed and will reopen a new PR so can get it tested.

**Pull request closed without merging**